proteus/mprans/SpatialTools.py

```diff
@@ -471,161 +473,159 @@ def setDimensions(self, dim):
-            self.regionIndice['y-'] = r_i
-            regionFlags += [r_i+1]
-            v_i += 2  # 2 vertices were added
-            r_i += 1  # 1 region was added
-            nb_sponge += 1
+            facets += [[[0, 1, v_i+1, v_i]],
+                       [[4, 5, v_i+3, v_i+2]],
+                       [[0, 1, 5, 4]],
+                       [[0, v_i, v_i+2, 4]],
+                       [[1, v_i+1, v_i+3, 5]]]
+            facetFlags += [bt['z-'],
+                           bt['z+'],
+                           bt['sponge']]
+            if x_n > 0:
+                facetFlags += [bt['wall']]
+            else:
+                facetFlags += [bt['x-']]
+            if x_p > 0:
+                facetFlags += [bt['wall']]
+            else:
+                facetFlags += [bt['x+']]
+            volumes[0][0][-1] = f_i+3 
+            volumes += [[[ f_i+i for i in range(6)]]]
+        v_i += 4  # 2 vertices were added
+        r_i += 1  # 1 region was added
+        nb_sponge += 1
+        f_i = len(facets)
+        # y+
+        vertices += [[x1, y1+y_p, z0], [x0, y1+y_p, z0],
+                     [x1, y1+y_p, z1], [x0, y1+y_p, z1]]
+        vertexFlags += [bt['z-'], bt['z-'],
+                        bt['z+'], bt['z+']]
+        volumes[0][0] += [f_i]  # add to volume
+        facets += [[[v_i, v_i+1, v_i+3, v_i+2]]]
+        facetFlags += [bt['y+']]
+        regions += [[(x0+x1)/2., (y1+(y1+y_p))/2., (z0+z1)/2.]]
+        self.regionIndice['y+'] = r_i
+        regionFlags += [r_i+1]
         if y_p:
-            vertices += [[x0, y1+y_p, z0], [x1, y1+y_p, z0]]
-            segments += [[3, v_i], [v_i, v_i+1], [v_i+1, 2]]
-            facets += [[[3, 2, v_i+1, v_i]]]
-            regions += [[(x0+x1)/2., (y1+(y1+y_p))/2., (z0+z1)/2.]]
-            self.regionIndice['y+'] = r_i
-            regionFlags += [r_i+1]
-            v_i += 2
-            r_i += 1
-            nb_sponge += 1
+            facets += [[[2, 3, v_i+1, v_i]],
+                       [[6, 7, v_i+3, v_i+2]],
+                       [[2, 3, 7, 6]],
+                       [[2, v_i, v_i+2, 6]],
+                       [[3, v_i+1, v_i+3, 7]]]
+            facetFlags += [bt['z-'],
+                           bt['z+'],
+                           bt['sponge']]
+            if x_p > 0:
+                facetFlags += [bt['wall']]
+            else:
+                facetFlags += [bt['x+']]
+            if x_n > 0:
+                facetFlags += [bt['wall']]
+            else:
+                facetFlags += [bt['x-']]
+            volumes[0][0][-1] = f_i+3 
+            volumes += [[[f_i+i for i in range(6)]]]
+        v_i += 4
+        r_i += 1
+        nb_sponge += 1
+        f_i = len(facets)
+        # x+
+        vertices += [[x1+x_p, y0, z0], [x1+x_p, y1, z0],
+                     [x1+x_p, y0, z1], [x1+x_p, y1, z1]]
+        vertexFlags += [bt['z-'], bt['z-'],
+                        bt['z+'], bt['z+']]
+        volumes[0][0] += [f_i]  # add to volume
```

Comment:
Could you explain me how volumes works?

Reply:
Volumes are only foto build a gmsh mesh, they are a facet loop basically